src/plugins/dist_matrixops/all_gather.cpp

```diff
@@ -0,0 +1,383 @@
+            local_result
+            Arg:
+
+                local_result (array) : a distributed array. A vector or matrix.
+
+            Returns:
+
+                A future holding a 2-D array with all values send
+                    by all participating localities.)"
+            )
+    };
+
+    ///////////////////////////////////////////////////////////////////////////
+    all_gather::all_gather(
+        execution_tree::primitive_arguments_type&& operands,
+        std::string const& name, std::string const& codename)
+      : primitive_component_base(std::move(operands), name, codename)
+    {}
+
+    ///////////////////////////////////////////////////////////////////////////
+    namespace detail
+    {
+        template <typename T>
+        execution_tree::primitive_argument_type concatenate2d_axis0(
+            execution_tree::primitive_arguments_type&& args,
+            std::string const& name, std::string const& codename)
+        {
+            std::size_t args_size = args.size();
+            std::array<std::size_t, PHYLANX_MAX_DIMENSIONS> prevdim =
+                extract_numeric_value_dimensions(args[0], name, codename);
+            std::size_t total_rows = 0;
+            for (std::size_t i = 0; i != args_size; ++i)
+            {
+                if (extract_numeric_value_dimension(args[i],
+                    name, codename) != 2)
+                {
+                    HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                        "dist_matrixops::primitives::all_gather::"
+                        "detail::concatenate2d_axis0",
+                        util::generate_error_message(
+                            "all the input arrays must have "
+                            "the same number of dimensions", name, codename));
+                }
+
+                std::array<std::size_t, PHYLANX_MAX_DIMENSIONS> dim =
+                    extract_numeric_value_dimensions(args[i], name, codename);
+
+                if (i != 0 && prevdim[1] != dim[1])
+                {
+                    HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                        "dist_matrixops::primitives::"
+                        "all_gather::detail::concatenate2d_axis0",
+                        util::generate_error_message(
+                            "all the input array dimensions except for "
+                            "the concatenation axis must match exactly ",
+                            name, codename));
+                }
+
+                total_rows += dim[0];
+                prevdim = dim;
+            }
+            blaze::DynamicMatrix<T> result(total_rows, prevdim[1]);
+            std::size_t step = 0;
+            for (auto&& arg : args)
+            {
+                auto&& val =
+                    execution_tree::extract_node_data<T>(std::move(arg));
+                std::size_t num_rows = val.dimension(0);
+                auto m = val.matrix();
+                for (std::size_t j = 0; j != num_rows; ++j)
```

Comment:
I think if we use a submatrix here we can get rid of the for loop.

Reply:
Yes. Will use submatrix instead.